src/components/Toolbar.js

```diff
@@ -1,6 +1,50 @@
 import React from 'react'
 import { connect } from 'react-redux'
 import { shortcutById } from '../shortcuts'
+import { overlayReveal, overlayHide, overlayUpdate } from '../action-creators/toolbar'
+import { CSSTransition, TransitionGroup } from 'react-transition-group'
+
+const isTouchEnabled = () => {
+    return ('ontouchstart' in window) ||
+           (navigator.maxTouchPoints > 0) ||
+           (navigator.msMaxTouchPoints > 0)
+}
+
+(() => {
+    window.addEventListener('mouseup', () => {
+        clearHoldTimer()
+    })
+    window.addEventListener('mouseover', e => {
+        const id = e.target.id
+        if (shortcutIds.includes(id)) {
+            updateOverlayInfo(id)
+        }
+    })
+    window.addEventListener('mousedown', e => {
+        const id = e.target.id
+        if (shortcutIds.includes(id)) {
+            clickOrHoldAction(id)
+        }
+    })
+    window.addEventListener('click', e => {
+        const id = e.target.id
+        if (shortcutIds.includes(id)) {
+            const { exec } = shortcutById(id)
+            exec(id)
+        }
+    })
+    if (isTouchEnabled()) {
+        window.addEventListener('touchstart', e => {
+            const id = e.target.id
+            if (shortcutIds.includes(id)) {
+                clickOrHoldAction(id)
+            }
+        })
```

Comment:
If there are any events that we can do on the element themselves rather than global, I think that would be better for performance. The `mouseup` and `touchend` events obviously need to be global due to the latest requirement.

Reply:
Got it.